cpp/src/arrow/c/abi.h

```diff
@@ -356,8 +356,7 @@ struct ArrowAsyncDeviceStreamHandler {
   // A producer that receives a non-zero return here should stop producing and eventually
   // call release instead.
   int (*on_schema)(struct ArrowAsyncDeviceStreamHandler* self,
-                   struct ArrowAsyncProducer* producer, struct ArrowSchema* stream_schema,
-                   const char* addl_metadata);
+                   struct ArrowSchema* stream_schema, const char* addl_metadata);
```

Comment:
If you already responded elsewhere then sorry for the noise, but I can't find my own first comment on this: we could remove the ambiguity of what this metadata pertains to by making it a member of `ArrowAsyncProducer`. This would simplify this signature and remove the need to clarify `addl_metadata`'s lifetime

Reply:
While it would simplify the signature I'm not sure I necessarily like the idea as far as the lifetime handling of it. 

@pitrou @lidavidm @paleolimbot what are your thoughts?

Reply:
The metadata might not be available until after the callback anyways so there'd instead be the question of when it's valid/invalid

Reply:
After `on_schema` all members of `self->producer` are initialized, right? So that would include `self->producer->metadata`. And all members of `self->producer` are expected to stay valid from then until `self` is destroyed, so that includes `self->producer->metadata`.

Reply:
I might be amenable to this idea, though I'm still unsure about forcing the keeping of the metadata around.

Using the suggestion from @bkietz that answers the question of lifetime and when it's valid/invalid. What do you think @lidavidm? 

this is the last question in play before we can merge this :)

Reply:
I suppose I don't mind then. As @kou notes let's spell it out too (`additional_metadata`)?

Reply:
Sounds good, I'll update this and the other two PRs accordingly

Reply:
I've updated this, shifting the `additional_metadata` member according to this suggestion. 

@bkietz if this is all good can I get an approval? I'll merge this after that.